File: embedding_matcher.py
#!/usr/bin/env python3
"""
EmbeddingMatcher: Semantic similarity matching for trigger/action names.

This module provides embedding-based matching as a fallback when pattern matching fails.
It uses a fine-tuned sentence-transformer model trained on Activepieces terminology.
"""

import json
import os
from pathlib import Path
from typing import List, Optional, Tuple, Dict
import numpy as np

class EmbeddingMatcher:
    """
    Semantic matcher using sentence embeddings.
    
    Falls back gracefully if the model is not available:
    - If sentence-transformers is not installed: returns None
    - If model is not trained: uses base model with reduced accuracy
    """
    
    MODEL_PATH = Path(__file__).parent / 'activepieces_embedding_model'
    BASE_MODEL = 'all-MiniLM-L6-v2'
    
    def __init__(self, threshold: float = 0.7):
        """
        Initialize the embedding matcher.
        
        Args:
            threshold: Minimum similarity score to consider a match (0-1)
        """
        self.threshold = threshold
        self.model = None
        self.available = False
        self._cache: Dict[str, np.ndarray] = {}
        
        self._load_model()
    
    def _load_model(self):
        """Load the sentence-transformer model."""
        try:
            from sentence_transformers import SentenceTransformer
            
            # Try fine-tuned model first
            if self.MODEL_PATH.exists():
                self.model = SentenceTransformer(str(self.MODEL_PATH))
                self.available = True
                print(f"  ℹ️  Loaded fine-tuned embedding model from {self.MODEL_PATH}")
            else:
                # Fall back to base model
                self.model = SentenceTransformer(self.BASE_MODEL)
                self.available = True
                print(f"  ⚠️  Fine-tuned model not found, using base model ({self.BASE_MODEL})")
                
        except ImportError:
            self.available = False
            # Silently fail - embedding matching is optional
    
    def _get_embedding(self, text: str) -> Optional[np.ndarray]:
        """Get embedding for a text, with caching."""
        if not self.available:
            return None
        
        if text not in self._cache:
            self._cache[text] = self.model.encode(text, convert_to_numpy=True)
        
        return self._cache[text]
    
    def _cosine_similarity(self, emb1: np.ndarray, emb2: np.ndarray) -> float:
        """Calculate cosine similarity between two embeddings."""
        return float(np.dot(emb1, emb2) / (np.linalg.norm(emb1) * np.linalg.norm(emb2)))
    
    def find_best_match(
        self, 
        query: str, 
        candidates: List[str],
        context: Optional[str] = None
    ) -> Optional[Tuple[str, float]]:
        """
        Find the best matching candidate for a query using semantic similarity.
        
        Args:
            query: The name to match (e.g., "notify_channel")
            candidates: List of valid names to match against
            context: Optional context (e.g., piece name) to improve matching
            
        Returns:
            Tuple of (best_match, score) if found, None if no match above threshold
        """
        if not self.available or not candidates:
            return None
        
        # Add context to query if provided
        if context:
            # Extract short piece name
            short_context = context.split('/')[-1].replace('piece-', '').replace('-', ' ')
            query_with_context = f"{short_context} {query}"
        else:
            query_with_context = query
        
        # Get query embedding
        query_emb = self._get_embedding(query_with_context)
        if query_emb is None:
            return None
        
        # Calculate similarity with each candidate
        best_match = None
        best_score = 0.0
        
        for candidate in candidates:
            # Also embed candidate with context for fair comparison
            if context:
                short_context = context.split('/')[-1].replace('piece-', '').replace('-', ' ')
                candidate_with_context = f"{short_context} {candidate}"
            else:
                candidate_with_context = candidate
            
            candidate_emb = self._get_embedding(candidate_with_context)
            if candidate_emb is None:
                continue
            
            similarity = self._cosine_similarity(query_emb, candidate_emb)
            
            if similarity > best_score:
                best_score = similarity
                best_match = candidate
        
        if best_match and best_score >= self.threshold:
            return (best_match, best_score)
        
        return None
    
    def find_matches(
        self,
        query: str,
        candidates: List[str],
        top_k: int = 3,
        context: Optional[str] = None
    ) -> List[Tuple[str, float]]:
        """
        Find top-k matching candidates for a query.
        
        Args:
            query: The name to match
            candidates: List of valid names
            top_k: Number of top matches to return
            context: Optional context for better matching
            
        Returns:
            List of (match, score) tuples, sorted by score descending
        """
        if not self.available or not candidates:
            return []
        
        # Add context
        if context:
            short_context = context.split('/')[-1].replace('piece-', '').replace('-', ' ')
            query_with_context = f"{short_context} {query}"
        else:
            query_with_context = query
        
        query_emb = self._get_embedding(query_with_context)
        if query_emb is None:
            return []
        
        # Calculate all similarities
        similarities = []
        for candidate in candidates:
            if context:
                candidate_with_context = f"{short_context} {candidate}"
            else:
                candidate_with_context = candidate
            
            candidate_emb = self._get_embedding(candidate_with_context)
            if candidate_emb is not None:
                similarity = self._cosine_similarity(query_emb, candidate_emb)
                similarities.append((candidate, similarity))
        
        # Sort by similarity and return top-k
        similarities.sort(key=lambda x: x[1], reverse=True)
        return similarities[:top_k]
    
    def is_available(self) -> bool:
        """Check if embedding matching is available."""
        return self.available


# Singleton instance for global use
_embedding_matcher: Optional[EmbeddingMatcher] = None

def get_embedding_matcher(threshold: float = 0.7) -> EmbeddingMatcher:
    """Get or create the global embedding matcher instance."""
    global _embedding_matcher
    if _embedding_matcher is None:
        _embedding_matcher = EmbeddingMatcher(threshold=threshold)
    return _embedding_matcher


def test_matcher():
    """Test the embedding matcher."""
    print("Testing EmbeddingMatcher...")
    
    matcher = EmbeddingMatcher(threshold=0.6)
    
    if not matcher.is_available():
        print("❌ Embedding matcher not available (missing dependencies)")
        return
    
    print("✅ Embedding matcher available\n")
    
    # Test cases
    test_cases = [
        {
            'query': 'notify_channel',
            'candidates': ['send_channel_message', 'send_direct_message', 'update_message', 'get_channel_history'],
            'context': '@activepieces/piece-slack',
            'expected': 'send_channel_message'
        },
        {
            'query': 'add_row',
            'candidates': ['insert_row', 'update_row', 'delete_row', 'find_rows'],
            'context': '@activepieces/piece-google-sheets',
            'expected': 'insert_row'
        },
        {
            'query': 'new_email',
            'candidates': ['gmail_new_email_received', 'send_email', 'gmail_get_mail'],
            'context': '@activepieces/piece-gmail',
            'expected': 'gmail_new_email_received'
        },
        {
            'query': 'receive_webhook',
            'candidates': ['catch_webhook'],
            'context': '@activepieces/piece-webhook',
            'expected': 'catch_webhook'
        },
        {
            'query': 'post_to_slack',
            'candidates': ['send_channel_message', 'send_direct_message', 'update_message'],
            'context': '@activepieces/piece-slack',
            'expected': 'send_channel_message'
        },
    ]
    
    passed = 0
    for tc in test_cases:
        result = matcher.find_best_match(tc['query'], tc['candidates'], tc['context'])
        if result:
            match, score = result
            status = "✅" if match == tc['expected'] else "❌"
            print(f"{status} '{tc['query']}' → '{match}' (score: {score:.3f}, expected: '{tc['expected']}')")
            if match == tc['expected']:
                passed += 1
        else:
            print(f"❌ '{tc['query']}' → No match (expected: '{tc['expected']}')")
    
    print(f"\nPassed: {passed}/{len(test_cases)}")


if __name__ == '__main__':
    test_matcher()

File: packages/server/api/src/app/flows/flow-post-processor/embedding-matcher.ts
/**
 * EmbeddingMatcher: Semantic similarity matching for trigger/action names.
 * 
 * Ported from Python to TypeScript using @xenova/transformers.
 * Provides embedding-based matching as a fallback when pattern matching fails.
 * 
 * Always loads the fine-tuned model from activepieces_embedding_model/ directory.
 * Throws an error if the model is not found.
 */

import type { FeatureExtractionPipeline } from '@xenova/transformers';
import path from 'path';

export interface MatchResult {
  name: string;
  score: number;
}

export class EmbeddingMatcher {
  private extractor: FeatureExtractionPipeline | null = null;
  private initialized: boolean = false;
  private initPromise: Promise<void> | null = null;
  private readonly threshold: number;
  private readonly cache: Map<string, Float32Array> = new Map();
  
  // Model path - always uses the fine-tuned model
  private readonly modelPath: string | null;
  
  constructor(threshold: number = 0.7, modelPath?: string) {
    this.threshold = threshold;
    // Always use the fine-tuned model from activepieces_embedding_model
    this.modelPath = modelPath || this.getLocalModelPath();
    
    if (!this.modelPath) {
      // This should rarely happen since the model is in the repo
      // But if it does, we'll handle it gracefully in init()
      console.warn(
        '⚠️  Fine-tuned embedding model not found at expected locations. ' +
        'Embedding matching will be disabled. Pattern matching will still work.'
      );
    }
  }
  
  private getLocalModelPath(): string | null {
    // Try to find the local fine-tuned model
    const possiblePaths = [
      path.join(process.cwd(), 'activepieces_embedding_model'),
      path.join(__dirname, '../../../../../activepieces_embedding_model'),
    ];
    
    for (const modelPath of possiblePaths) {
      try {
        const fs = require('fs');
        if (fs.existsSync(modelPath)) {
          // Verify it's a valid model directory
          const configPath = path.join(modelPath, 'config.json');
          if (fs.existsSync(configPath)) {
            return modelPath;
          }
        }
      } catch {
        // Continue to next path
      }
    }
    
    return null;
  }
  
  /**
   * Initialize the embedding model (lazy loading).
   */
  private async init(): Promise<void> {
    if (this.initialized || this.initPromise) {
      if (this.initPromise) {
        await this.initPromise;
      }
      return;
    }
    
    this.initPromise = (async () => {
      if (!this.modelPath) {
        // Model path not found, disable embeddings
        this.extractor = null;
        this.initialized = true;
        return;
      }
      
      try {
        // Dynamically import transformers to avoid bundling issues
        // NOTE: This is only used for local TypeScript post-processing
        // When using the Python model server, embeddings are handled there
        const transformers = await import('@xenova/transformers');
        
        // Use feature-extraction pipeline for embeddings
        this.extractor = await transformers.pipeline(
          'feature-extraction',
          this.modelPath!,
          {
            quantized: false, // Use full precision for better accuracy
            progress_callback: (progress: any) => {
              if (progress.status === 'loading') {
                console.log(`  ℹ️  Loading embedding model: ${progress.file || ''}`);
              }
            },
          }
        ) as FeatureExtractionPipeline;
        
        this.initialized = true;
        console.log(`  ✅ Embedding matcher initialized (model: ${this.modelPath})`);
      } catch (error) {
        console.warn(`  ⚠️  Failed to load embedding model: ${error}`);
        this.extractor = null;
        this.initialized = true; // Mark as initialized to prevent retries
      }
    })();
    
    await this.initPromise;
  }
  
  /**
   * Get embedding for a text, with caching.
   */
  private async getEmbedding(text: string): Promise<Float32Array | null> {
    if (!this.extractor) {
      await this.init();
      if (!this.extractor) {
        return null;
      }
    }
    
    // Check cache
    if (this.cache.has(text)) {
      return this.cache.get(text)!;
    }
    
    try {
      // Get embedding
      const output = await this.extractor(text, {
        pooling: 'mean',
        normalize: true,
      });
      
      // Convert to Float32Array
      const embedding = Array.isArray(output) 
        ? new Float32Array(output.flat())
        : new Float32Array(output.data);
      
      // Cache it
      this.cache.set(text, embedding);
      
      return embedding;
    } catch (error) {
      console.warn(`  ⚠️  Failed to get embedding for "${text}": ${error}`);
      return null;
    }
  }
  
  /**
   * Calculate cosine similarity between two embeddings.
   */
  private cosineSimilarity(emb1: Float32Array, emb2: Float32Array): number {
    if (emb1.length !== emb2.length) {
      return 0;
    }
    
    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;
    
    for (let i = 0; i < emb1.length; i++) {
      dotProduct += emb1[i] * emb2[i];
      norm1 += emb1[i] * emb1[i];
      norm2 += emb2[i] * emb2[i];
    }
    
    const denominator = Math.sqrt(norm1) * Math.sqrt(norm2);
    if (denominator === 0) {
      return 0;
    }
    
    return dotProduct / denominator;
  }
  
  /**
   * Find the best matching candidate for a query using semantic similarity.
   * 
   * @param query - The name to match (e.g., "notify_channel")
   * @param candidates - List of valid names to match against
   * @param context - Optional context (e.g., piece name) to improve matching
   * @returns Tuple of [best_match, score] if found, null if no match above threshold
   */
  async findBestMatch(
    query: string,
    candidates: string[],
    context?: string
  ): Promise<[string, number] | null> {
    if (!candidates || candidates.length === 0) {
      return null;
    }
    
    await this.init();
    
    if (!this.extractor) {
      return null;
    }
    
    // Add context to query if provided
    const queryWithContext = context
      ? this.buildContextualQuery(query, context)
      : query;
    
    // Get query embedding
    const queryEmb = await this.getEmbedding(queryWithContext);
    if (!queryEmb) {
      return null;
    }
    
    // Calculate similarity with each candidate
    let bestMatch: string | null = null;
    let bestScore = 0;
    
    for (const candidate of candidates) {
      // Also embed candidate with context for fair comparison
      const candidateWithContext = context
        ? this.buildContextualQuery(candidate, context)
        : candidate;
      
      const candidateEmb = await this.getEmbedding(candidateWithContext);
      if (!candidateEmb) {
        continue;
      }
      
      const similarity = this.cosineSimilarity(queryEmb, candidateEmb);
      
      if (similarity > bestScore) {
        bestScore = similarity;
        bestMatch = candidate;
      }
    }
    
    if (bestMatch && bestScore >= this.threshold) {
      return [bestMatch, bestScore];
    }
    
    return null;
  }
  
  /**
   * Find top-k matching candidates for a query.
   */
  async findMatches(
    query: string,
    candidates: string[],
    topK: number = 3,
    context?: string
  ): Promise<MatchResult[]> {
    if (!candidates || candidates.length === 0) {
      return [];
    }
    
    await this.init();
    
    if (!this.extractor) {
      return [];
    }
    
    const queryWithContext = context
      ? this.buildContextualQuery(query, context)
      : query;
    
    const queryEmb = await this.getEmbedding(queryWithContext);
    if (!queryEmb) {
      return [];
    }
    
    // Calculate all similarities
    const similarities: MatchResult[] = [];
    
    for (const candidate of candidates) {
      const candidateWithContext = context
        ? this.buildContextualQuery(candidate, context)
        : candidate;
      
      const candidateEmb = await this.getEmbedding(candidateWithContext);
      if (candidateEmb) {
        const similarity = this.cosineSimilarity(queryEmb, candidateEmb);
        similarities.push({
          name: candidate,
          score: similarity,
        });
      }
    }
    
    // Sort by similarity and return top-k
    similarities.sort((a, b) => b.score - a.score);
    return similarities.slice(0, topK);
  }
  
  /**
   * Build a contextual query by adding piece name context.
   */
  private buildContextualQuery(name: string, context: string): string {
    // Extract short piece name (e.g., "piece-slack" from "@activepieces/piece-slack")
    const shortContext = context
      .split('/')
      .pop()
      ?.replace('piece-', '')
      .replace(/-/g, ' ') || '';
    
    return `${shortContext} ${name}`;
  }
  
  /**
   * Check if embedding matching is available.
   */
  isAvailable(): boolean {
    return this.extractor !== null;
  }
  
  /**
   * Clear the embedding cache.
   */
  clearCache(): void {
    this.cache.clear();
  }
}

// Singleton instance for global use
let embeddingMatcherInstance: EmbeddingMatcher | null = null;

export function getEmbeddingMatcher(threshold: number = 0.7): EmbeddingMatcher {
  if (!embeddingMatcherInstance) {
    embeddingMatcherInstance = new EmbeddingMatcher(threshold);
  }
  return embeddingMatcherInstance;
}

